### PR TITLE
Exposing additional events

### PR DESCRIPTION
**Pull request by seanfarmar**
https://github.com/Particular/ServiceControl/issues/1445

### CODE REVIEW

File: src/ServiceControl/ExternalIntegrations/FailedMessagesUnArchivedPublisher.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+                MessagesCount = @event.MessagesCount
+            };
+        }
+
+        protected override Task<IEnumerable<object>> PublishEvents(IEnumerable<DispatchContext> contexts, IAsyncDocumentSession session)
+        {
+            return Task.FromResult(contexts.Select(r => (object)new Contracts.FailedMessagesUnArchived
+            {
+                MessagesCount = r.MessagesCount,
+                DocumentIds = r.MessageIds
```
Comment (danielmarbach) [inline src/ServiceControl/ExternalIntegrations/FailedMessagesUnArchivedPublisher.cs:25]: Why are we publishing DocumentIds? How is that relevant for someone observing this event that is external to RavenDB? Shouldn't this be called MessagesIds?

File: src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesByRangeHandler.cs [hunk 2]
```diff
@@ -50,6 +51,7 @@ public async Task Handle(UnArchiveMessagesByRange message, IMessageHandlerContex
 
             await domainEvents.Raise(new FailedMessagesUnArchived
             {
+                DocumentIds = patchedDocumentIds.Select(x => x.Document).ToArray(),
```
Comment (danielmarbach) [inline src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesByRangeHandler.cs:54]: Is this even correct? Aren't we patching against `FailedMessageViewIndex` and that index is using `FailedMessage` where the `Id` is determined by

```
        public static string MakeDocumentId(string messageUniqueId)
        {
            return $"{CollectionName}/{messageUniqueId}";
        }
```

which then would mean we'd have to do something like


```
DocumentIds = patchedDocumentIds.Select(x => FailedMessage.GetMessageIdFromDocumentId(x.Document)).ToArray(),
```

?

File: src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                FailedMessageId = new Guid(@event.FailedMessageId)
+            };
+        }
+
+        protected override Task<IEnumerable<object>> PublishEvents(IEnumerable<DispatchContext> contexts, IAsyncDocumentSession session)
+        {
+            return Task.FromResult(contexts.Select(r => (object)new Contracts.MessageFailureResolvedByRetry
+            {
+                FailedMessageId = r.FailedMessageId.ToString(),
+                AlternativeFailedMessageIds = r.AlternativeFailedMessageIds
```
Comment (danielmarbach) [inline src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs:25]: Isn't this information always empty? We have `RetryHandlerForBackwardsCompatability` and it doesn't set the array. Ah we set it in the recoverability specific enricher in the audit pipeline. 

```
            IEnumerable<string> GetAlternativeUniqueMessageId(IReadOnlyDictionary<string, string> headers)
            {
                var messageId = headers.MessageId();
                if (headers.TryGetValue(Headers.ProcessingEndpoint, out var processingEndpoint))
                {
                    yield return DeterministicGuid.MakeId(messageId, processingEndpoint).ToString();
                }

                if (headers.TryGetValue("NServiceBus.FailedQ", out var failedQ))
                {
                    yield return DeterministicGuid.MakeId(messageId, ExtractQueueNameForLegacyReasons(failedQ)).ToString();
                }

                if (headers.TryGetValue(Headers.ReplyToAddress, out var replyToAddress))
                {
                    yield return DeterministicGuid.MakeId(messageId, ExtractQueueNameForLegacyReasons(replyToAddress)).ToString();
                }
            }
```

How is that information relevant for the end users? Can you elaborate what they would do with this information?

File: src/ServiceControl/ExternalIntegrations/FailedMessagesUnArchivedPublisher.cs (hunk 1 above)
Comment (seanfarmar) [reply src/ServiceControl/ExternalIntegrations/FailedMessagesUnArchivedPublisher.cs:25]: Make sense, I'll change the contract

File: src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs (hunk 3 above)
Comment (seanfarmar) [reply src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs:25]: In the case I was looking at the customer wants to use the events to reconcile his backup error collection, so he wants to know if he needs to remove a failed message from his backup. (so if it was resolved, archived or unarchived)

**Comment by danielmarbach** [comment]
I think this lacks also acceptance tests under the corresponding External integration folder. See for example `When_a_message_has_failed_detected`

File: src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    using Contracts.MessageFailures;
+    using Raven.Client;
+
+    class MessageFailureResolvedByRetryPublisher : EventPublisher<MessageFailureResolvedByRetry, MessageFailureResolvedByRetryPublisher.DispatchContext>
+    {
+        protected override DispatchContext CreateDispatchRequest(MessageFailureResolvedByRetry @event)
+        {
+            return new DispatchContext
+            {
+                FailedMessageId = new Guid(@event.FailedMessageId)
```
Comment (mikeminutillo) [inline src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs:16]: Was there a reason to convert this to a Guid only to then convert it back to a string when we publish it? Raven is likely going to store it as a json string anyway.

File: src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesByRangeHandler.cs [hunk 5]
```diff
@@ -50,6 +51,7 @@ public async Task Handle(UnArchiveMessagesByRange message, IMessageHandlerContex
 
             await domainEvents.Raise(new FailedMessagesUnArchived
             {
+                DocumentIds = patchedDocumentIds.Select(x => FailedMessage.GetMessageIdFromDocumentId(x.Document)).ToArray(),
```
Comment (mikeminutillo) [inline src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesByRangeHandler.cs:54]: Is there a limit to how many of these there can be? We may want to batch them otherwise these messages get very large.

File: src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesHandler.cs [hunk 6]
```diff
@@ -41,6 +41,7 @@ await session.SaveChangesAsync()
 
             await domainEvents.Raise(new FailedMessagesUnArchived
             {
+                DocumentIds = failedMessages.Select(x => x.UniqueMessageId).ToArray(),
```
Comment (mikeminutillo) [inline src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesHandler.cs:44]: We may want to batch these to keep the message sizes low. I am not sure if there are natural limits to how big this collection can be.

**Review by mikeminutillo — COMMENTED** [review]
Probably also need to expose `FailedMessageGroupArchived`.

Do we know what the impact on performance is?

For Azure Service Bus this will increase the number of operations against the broker, even if there are no subscribers. Should we consider having a way to disable some/all integration events?

File: src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs (hunk 4 above)
Comment (seanfarmar) [reply src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs:16]: Not that I can think of, I'll check if I can leave it as string :-)

File: src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesHandler.cs (hunk 6 above)
Comment (seanfarmar) [reply src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesHandler.cs:44]: What kind of batch size do you think we should do here?
Is there a metric we can use?

File: src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesByRangeHandler.cs (hunk 5 above)
Comment (seanfarmar) [reply src/ServiceControl/MessageFailures/Handlers/UnArchiveMessagesByRangeHandler.cs:54]: What kind of batch size do you think we should do here?
Is there a metric we can use?

File: src/ServiceControl.Audit/ServiceControl.Audit.csproj [hunk 7]
```diff
@@ -33,7 +33,7 @@
     <PackageReference Include="NServiceBus.NLog" Version="3.0.0" />
     <PackageReference Include="NServiceBus.Raw" Version="3.2.2" />
     <PackageReference Include="Microsoft.Owin.Hosting" Version="4.1.1" />
-    <PackageReference Include="ServiceControl.Contracts" Version="2.0.0" />
+    <PackageReference Include="ServiceControl.Contracts" Version="3.0.0-PullRequest0046.9" />
```
Comment (ramonsmits) [inline src/ServiceControl.Audit/ServiceControl.Audit.csproj:36]: Incorrect package

File: src/ServiceControl/ServiceControl.csproj [hunk 8]
```diff
@@ -34,7 +34,7 @@
     <PackageReference Include="NServiceBus.NLog" Version="3.0.0" />
     <PackageReference Include="NServiceBus.Raw" Version="3.2.2" />
     <PackageReference Include="Microsoft.Owin.Hosting" Version="4.1.1" />
-    <PackageReference Include="ServiceControl.Contracts" Version="2.0.0" />
+    <PackageReference Include="ServiceControl.Contracts" Version="3.0.0-PullRequest0046.9" />
```
Comment (ramonsmits) [inline src/ServiceControl/ServiceControl.csproj:37]: Same here, incorrect package.

File: src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt [hunk 9]
```diff
@@ -1174,6 +1175,7 @@ namespace ServiceControl.Recoverability
     public class FailedMessageGroupArchived : ServiceControl.Infrastructure.SignalR.IUserInterfaceEvent
     {
         public FailedMessageGroupArchived() { }
+        public string[] FailedMessagesIds { get; set; }
```
Comment (ramonsmits) [inline src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt:1178]: How many IDs can be in this array? Any issues with transport message sizes? I don't see any bound checks so wondering what the max size is.

As this is text and often van contain GUID that is 36 characters, with quotes and comma its 39. Meaning, 128kb can contain roughly ~3.000 ID's assuming these are GUID's.

**Review by ramonsmits — CHANGES_REQUESTED** [review]

File: src/ServiceControl.Audit/ServiceControl.Audit.csproj (hunk 7 above)
Comment (seanfarmar) [reply src/ServiceControl.Audit/ServiceControl.Audit.csproj:36]: It is the new contracts, it will be updated once we release it

File: src/ServiceControl/ServiceControl.csproj (hunk 8 above)
Comment (seanfarmar) [reply src/ServiceControl/ServiceControl.csproj:37]: It is the new contracts, it will be updated once we release it

File: src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt (hunk 9 above)
Comment (SzymonPobiega) [reply src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt:1178]: Good catch @ramonsmits . We have added it to a number of contracts. If the number of items was bigger than the max allowed message size the publish would fail and the event would not be delivered. Given that we don't have a good way of checking it I'd say we ignore it for now and document architecture cruft here. 

Alternatively we could split one domain event into multiple on-the-wire events with portions of the data.

Comment (seanfarmar) [reply src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt:1178]: Good point, Do we any pattern/guidance about that? Does that depend on the transport we use?

Comment (SzymonPobiega) [reply src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt:1178]: Unfortunately depends on the transport :\

Comment (tmasternak) [reply src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt:1178]: We could add a new domain event when [creating retry batches](https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/Recoverability/Archiving/ArchiveAllInGroupHandler.cs#L66) and publish an integration event with message ids in that batch.

@seanfarmar do we need `GroupId` and `GroupName` in some of the integration events?

Comment (seanfarmar) [reply src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.PublicClr.approved.txt:1178]: @tmasternak For the scenario I'm thinking of, we don't because the customer is only tacking single messages...

File: src/ServiceControl/Recoverability/Archiving/Raven/ArchiveOperation.cs [hunk 10]
```diff
@@ -13,6 +14,7 @@ class ArchiveOperation // raven
         public DateTime Started { get; set; }
         public int NumberOfBatches { get; set; }
         public int CurrentBatch { get; set; }
+        public List<string> FailedMessagesIds { get; set; }
```
Comment (SzymonPobiega) [inline src/ServiceControl/Recoverability/Archiving/Raven/ArchiveOperation.cs:17]: I think we don't need this one any more.

File: src/ServiceControl/ExternalIntegrations/FailedMessageArchivedPublisher.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            return Task.FromResult(contexts.Select(r => (object)new Contracts.FailedMessagesArchived
+            {
+                FailedMessagesIds = new[] { r.FailedMessageId.ToString() }
+            }));
+        }
+
+        public class DispatchContext
+        {
+            public Guid FailedMessageId { get; set; }
```
Comment (SzymonPobiega) [inline src/ServiceControl/ExternalIntegrations/FailedMessageArchivedPublisher.cs:30]: We can replace this one by `string FailedMessageId`. Both are serialized to string in json so there should be no problem with deserialization and it will save us the conversions in lines 16 and 24.

File: src/ServiceControl/ExternalIntegrations/FailedMessageGroupArchivedPublisher.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+                FailedMessagesIds = @event.FailedMessagesIds
+            };
+        }
+
+        protected override Task<IEnumerable<object>> PublishEvents(IEnumerable<DispatchContext> contexts, IAsyncDocumentSession session)
+        {
+            return Task.FromResult(contexts.Select(r => (object)new Contracts.FailedMessagesArchived
+            {
+                // cleanup FailedMessages/ publish guids without document collection name
+                FailedMessagesIds = r.FailedMessagesIds.Select(id => id.Replace("FailedMessages/", "")).ToArray()
```
Comment (SzymonPobiega) [inline src/ServiceControl/ExternalIntegrations/FailedMessageGroupArchivedPublisher.cs:24]: How about we remove the `FailedMessages` part when publishing the domain event so that we don't need to do it here? I think that would be cleaner and make it align with how `FailedMessageArchivedPublisher` and `FailedMessagesUnArchivedPublisher` work.

**Review by SzymonPobiega — DISMISSED** [review]
Minor suggestions

**Review by SzymonPobiega — APPROVED** [review]

File: src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_archived.cs [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    {
+        [Test]
+        public async Task Should_publish_notification()
+        {
+            CustomConfiguration = config => config.OnEndpointSubscribed<MyContext>((s, ctx) =>
+            {
+                if (s.SubscriberReturnAddress.IndexOf("ExternalProcessor", StringComparison.OrdinalIgnoreCase) >= 0)
+                {
+                    ctx.ExternalProcessorSubscribed = true;
+                }
```
Comment (tmasternak) [inline src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_archived.cs:28]:
```suggestion
                ctx.ExternalProcessorSubscribed = s.SubscriberReturnAddress.Contains("ExternalProcessor");
```

File: src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_resolved_by_retry.cs [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        [Test]
+        public async Task Should_publish_notification()
+        {
+            CustomConfiguration = config => config.OnEndpointSubscribed<MyContext>((s, ctx) =>
+            {
+                if (s.SubscriberReturnAddress.IndexOf("ExternalProcessor", StringComparison.OrdinalIgnoreCase) >= 0)
+                {
+                    ctx.ExternalProcessorSubscribed = true;
+                }
```
Comment (tmasternak) [inline src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_resolved_by_retry.cs:28]:
```suggestion
                ctx.ExternalProcessorSubscribed = s.SubscriberReturnAddress.Contains("ExternalProcessor");
```

File: src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_resolved_by_retry.cs [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+                        e => e.Status == FailedMessageStatus.Resolved);
+                })
+                .Done(ctx => ctx.EventDelivered) //Done when sequence is finished
+                .Run();
+
+            var deserializedEvent = JsonConvert.DeserializeObject<MessageFailureResolvedByRetry>(context.Event);
+            Assert.IsTrue(deserializedEvent.FailedMessageId == context.FailedMessageId);
+        }
+
+        public class Receiver : EndpointConfigurationBuilder
```
Comment (tmasternak) [inline src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_resolved_by_retry.cs:79]: Could we move this to a test base class to prevent duplication?

File: src/ServiceControl/Recoverability/Archiving/ArchiveAllInGroupHandler.cs [hunk 16]
```diff
@@ -89,6 +91,15 @@ await documentManager.UpdateArchiveOperation(batchSession, archiveOperation)
                     await batchSession.SaveChangesAsync()
                         .ConfigureAwait(false);
 
+                    if (nextBatch != null)
+                    {
+                        await domainEvents.Raise(new FailedMessageGroupBatchArchived
+                        {
+                            // cleanup FailedMessages/ publish guids without document collection name
```
Comment (tmasternak) [inline src/ServiceControl/Recoverability/Archiving/ArchiveAllInGroupHandler.cs:98]:
```suggestion
                            // Remove `FailedMessages/` prefix and publish pure GUIDs without Raven collection name
```

**Review by tmasternak — CHANGES_REQUESTED** [review]

File: src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_resolved_by_retry.cs (hunk 15 above)
Comment (seanfarmar) [reply src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_resolved_by_retry.cs:79]: @tmasternak you are refactoring this now, so ignoring this for now

File: src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_archived.cs (hunk 13 above)
Comment (mikeminutillo) [reply src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_archived.cs:28]: Technically speaking this suggestion could reset the flag to false if a second subscriber exists.

**Review by mikeminutillo — COMMENTED** [review]
All in all this looks good. My chief concerns would be:

- Publishing the results of bulk opertaions with a list of GUIDs could quickly create very large messages. We should maybe consider batching these at the integration event level.
- Publishing more events may incur more costs in cloud environments. We should look at providing options to disable some/all of these event types.

**Comment by SzymonPobiega** [comment]
> Publishing the results of bulk opertaions with a list of GUIDs could quickly create very large messages. We should maybe consider batching these at the integration event level.

We have changed the code to publish these messages for each archived batch (1000 docs) so we are safe here that we won't cross the threshold even on the most restricted transport (ASQ)

**Comment by SzymonPobiega** [comment]
> We should maybe consider batching these at the integration event level

I think we are not making things significantly worse because the biggest impact on cost is probably `MessageFailed` event.

**Comment by mikeminutillo** [comment]
> > We should maybe consider batching these at the integration event level
> 
> I think we are not making things significantly worse because the biggest impact on cost is probably `MessageFailed` event.

My concern is that for every `MessageFailed` event we are adding at least one more eventual event. So the number of events published will be 2x.

**Comment by tmasternak** [comment]
> My concern is that for every MessageFailed event we are adding at least one more eventual event. So the number of events published will be 2x.

One more event is a situation that was already there before we added new events. Secondly, these events won't get published unless there is an integration event subscriber in the system.

**Comment by mikeminutillo** [comment]
> One more event is a situation that was already there before we added new events. 

I am not following this. Can you explain how this situation was already present?

> Secondly, these events won't get published unless there is an integration event subscriber in the system.

For message-driven pub-sub that is true. For native-pub sub transports like Azure Service Bus (where every operation costs money) we still are going to push the event to the topic, even if there are no subscribers for it.

**Comment by seanfarmar** [comment]
From Slack https://particularsoftware.slack.com/archives/C01SQ08L5S8/p1618993470145600

@mikeminutillo's summary:

To summarize the conversation:
- We already publish 1 event for every failed message
- Now, if you retry, we will publish 1 additional event for every failed message
- If you archive, then we batch those notification events
- If every message is retried once, we double the number of events published
- On most transports, this won't matter much
  - For any with some kind of subscription storage (native or message-driven), we will see that there are no subscribers for these new messages and not bother publishing them
  - For any with native pub-sub, we will publish the message to the broker regardless. This puts more pressure on the broker.
    - For RabbitMQ that's probably OK. Rabbit is often much faster than we are anyway and the customer has already paid the cost of setting up the infra. There is usually not an additional cost per operation.
    - For Azure Service Bus that's probably not OK. Each operation costs money and we may be publishing events that the customer doesn't even care about. For them, they will upgrade to a new version and their costs will go up. They will  have questions about that.

The outcome is that we will introduce a config flag to turn off the publishing of integration events.
- This flag allows those ASB customers to switch off these events if they don't use them.
  - Customers that don't care about any integration events will benefit from this, and we might be able to reduce costs for them
  - There may be a small number of customers that do care about the old integration events but not the new ones. I think we'd like to hear from them before we build anything more complex
- Even non-ASB customers might be interested in this because it takes the pressure off of RavenDB.
  - In high-throughput scenarios where we don't use Integration Events, I would advise turning this feature off.
  - I'd even go so far as to make it off by default, but I'd want to do that in a major with a user experience around turning it back on at runtime

File: src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_archived.cs [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    using NServiceBus;
+    using NServiceBus.AcceptanceTesting;
+    using NUnit.Framework;
+    using Contracts;
+    using ServiceControl.MessageFailures;
+    using Newtonsoft.Json;
+    using TestSupport.EndpointTemplates;
+    using ServiceBus.Management.Infrastructure.Settings;
+
+    class When_a_failed_message_is_archived : When_a_message_failed
```
Comment (danielmarbach) [inline src/ServiceControl.AcceptanceTests/Recoverability/ExternalIntegration/When_a_failed_message_is_archived.cs:14]: Is it possible to rename the base class to not contain `When_`? It know this sounds super nitpicky but for my brain a `When_` is always a concrete scenario but here it seems `When_a_message_failed` is just a convenience base class that provides infrastructure to avoid having to repeat things for external integration. If the fact that all those events are based upon message failure how about calling it something like 

- `MessageFailedAcceptanceTests` or
- `MessageFailedIntegrationAcceptanceTests`

File: src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+            return Task.FromResult(contexts.Select(r => (object)new Contracts.MessageFailureResolvedByRetry
+            {
+                FailedMessageId = r.FailedMessageId.ToString(),
+                AlternativeFailedMessageIds = r.AlternativeFailedMessageIds
+            }));
+        }
+
+        public class DispatchContext
+        {
+            public Guid FailedMessageId { get; set; }
```
Comment (danielmarbach) [inline src/ServiceControl/ExternalIntegrations/MessageFailureResolvedByRetryPublisher.cs:31]: Wouldn't it be simpler to just make this a string? Then you wouldn't have to convert to guid and back to string right?

File: src/ServiceControl/ExternalIntegrations/MessageFailureResolvedManuallyPublisher.cs [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            return Task.FromResult(contexts.Select(r => (object)new Contracts.MessageFailureResolvedManually
+            {
+                FailedMessageId = r.FailedMessageId.ToString()
+            }));
+        }
+
+        public class DispatchContext
+        {
+            public Guid FailedMessageId { get; set; }
```
Comment (danielmarbach) [inline src/ServiceControl/ExternalIntegrations/MessageFailureResolvedManuallyPublisher.cs:30]: Wouldn't it be simpler to just make this a string? Then you wouldn't have to convert to guid and back to string right?

**Review by danielmarbach — CHANGES_REQUESTED** [review]
A few comments

**Comment by tmasternak** [comment]
@danielmarbach can you have a look now? Thx.

**Review by danielmarbach — APPROVED** [review]
